### DISCUSSION TITLE
Accessing key name of parent

### DISCUSSION THREAD

**Discussion by jzazo** [omry/omegaconf#937] [Q&A] [answered]
Hi! I would like to access the parent key name of an attribute, probably via a resolver, would that be possible?
For example:
```
cfg = OmegaConf.create({"john": {"height": 180, "weight": 75, "player": "${parent_name:}"}})
cfg.john.player == "john"  # True
```
The reason is that I would like to pass the name of a parent to its children, e.g., a named storage that takes the name of the task that uses that storage system. Thank you!

**Comment by Jasha10** [comment, marked as answer]
Hi @jzazo!
I hope you don't mind using undocumented/internal OmegaConf methods, because I don't think this can be achieved with OmegaConf's public API.

Custom resolvers can take a special `_parent_` argument (this part *is* [documented](https://omegaconf.readthedocs.io/en/latest/custom_resolvers.html), just before the "Built in Resolvers" section). This, in combination with the undocumented `_key()` method, will be enough to do the job:
```python
from omegaconf import OmegaConf

def parent_name_impl(_parent_) -> str:
    return _parent_._key()

OmegaConf.register_new_resolver("parent_name", parent_name_impl)

cfg = OmegaConf.create({"john": {"height": 180, "weight": 75, "player": "${parent_name:}"}})
assert cfg.john.player == "john"
```
The standard caveats for undocumented/internal methods apply...

**Reply by jzazo** [reply]
Thanks a lot, your suggestion works great. I was wondering (for my use case), if it would be possible to get arbitrary grandparent name? Something like, given a node using relative interpolation, grab its parent, then grab its key? Would that be possible?

**Reply by jzazo** [reply]
I was thinking quite a convoluted solution: compute the path between the `_root_` and the `_parent_` of a node using dynamic programming, then once you have the path, grab the `_key()` at any level you want. Can you think of a better solution?

**Reply by Jasha10** [reply]
> Can you think of a better solution?

I can, but my solution depends on OmegaConf's internals. Whether this will work for you depends on if you are willing to tie your project to OmegaConf's implementation details (which might change in the future!)

```python
def grandparent_name_impl(_parent_) -> str:
    return _parent_._get_parent()._key()
```

See also [`Node._get_full_key`](https://github.com/omry/omegaconf/blob/0b0b38e0dd0e5c5c06ab8ff6b3440941831c078d/omegaconf/base.py#L242), which returns a string representation of the full path from `_root_` to the given node.

In the future I'll be working to extend [OmegaConf's `Unions` feature](https://omegaconf.readthedocs.io/en/latest/structured_config.html#unions) to support unions of non-primitive types (e.g. types hints like `Union[Dict[...], List[...]]`. Once this is implemented, and if you use the unions feature, it's possible you'll find that `Node._key()` returns `None` for some container nodes (specifically if the node is wrapped inside a `UnionNode`).

**Reply by jzazo** [reply]
Thanks again, these proposals work and are cleaner. I think I would not use them with the `Union`s that you mention, but I will keep an eye for errors if I get `None`s.
